srcts/src/components/darkModeSwitch.ts

```diff
@@ -0,0 +1,323 @@
+      }
+    `,
+    css`
+      :host {
+        display: inline-block;
+
+        /* We control the stroke size manually here. We don't want it getting so
+        small its not visible but also not so big it looks cartoonish */
+        --stroke-w: clamp(1px, 0.1em, 6px);
+      }
+
+      button {
+        /* This is needed to let the svg use the em sizes */
+        font-size: inherit;
+
+        /* Make sure the button is fully centered */
+        display: grid;
+        place-content: center;
+
+        /* A little bit of padding to make it easier to press */
+        padding: var(--size-xxs);
+        background: none;
+        border: none;
+        aspect-ratio: 1;
+        border-radius: 50%;
+        cursor: pointer;
+        touch-action: manipulation;
+        -webkit-tap-highlight-color: transparent;
+        outline-offset: var(--size-xxs);
+
+        /* Move down to adjust for being larger than 1em */
+        transform: translateY(var(--vertical-correction));
+        margin-block-end: var(--vertical-correction);
+      }
+
+      /*
+      button:is(:hover, :focus-visible) {
+        background: var(--surface-4);
+      }
+      */
+
+      button > svg {
+        height: var(--size);
+        width: var(--size);
+        stroke-linecap: round;
+        overflow: visible;
+      }
+
+      svg line,
+      svg circle {
+        vector-effect: non-scaling-stroke;
+      }
+    `,
+  ];
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
+  onChangeCallback: (x: boolean) => void = (x: boolean) => {};
+
+  // onValueChange = makeValueChangeEmitter(this, this.id);
+
+  connectedCallback(): void {
+    super.connectedCallback();
+
+    this.attribute = this.getAttribute("attribute") || "data-shinytheme";
+
+    if (typeof this.mode === "undefined") {
+      this.mode = window.matchMedia("(prefers-color-scheme: dark)").matches
+        ? "dark"
+        : "light";
+    }
+
+    this.reflectPreference();
+
+    // Sync with system changes
+    window
+      .matchMedia("(prefers-color-scheme: dark)")
+      .addEventListener("change", ({ matches: isDark }) => {
+        this.mode = isDark ? "dark" : "light";
+        this.reflectPreference();
+      });
+
+    this._observeDocumentThemeAttribute();
+  }
+
+  private _observeDocumentThemeAttribute(): void {
+    const observer = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        if (mutation.target !== document.documentElement) return;
+        if (mutation.attributeName !== this.attribute) return;
+        const newValue = document.documentElement.getAttribute(this.attribute);
+        if (!newValue || newValue === this.mode) return;
+        this.mode = newValue as DarkModeSwitch["mode"];
+      });
+    });
+
+    const config = {
+      attributes: true,
+      childList: false,
+      subtree: false,
+    };
+
+    observer.observe(document.documentElement, config);
```

Comment:
I just realized that this will leak MutationObservers and old DOM elements. For example, if the switch is added via `renderUI`, then each one will add a MutationObserver attached to `document.documentElement`, and each one will keep a reference to the old `BslibInputDarkMode` component, preventing it from getting GC'd. This callback will also run once for each instance of a `BslibInputDarkMode` that has ever appeared on the page.

I know this doesn't sound like it'll be a big problem in practice, but it'll be good to get this right so that we have something solid to refer to in the future. (And it actually could become an issue in practice, if someone makes a single-page app where the dynamically rendered content includes this component.)

Here's an example app that demonstrates the problem.

```
library(shiny)
pkgload::load_all()

ui <- page_fluid(
  uiOutput("ui"),
  actionButton("redraw", "Redraw UI")
)

server <- function(input, output, session) {
  count <- 0
  
  output$ui <- renderUI({
    input$redraw
    count <<- count + 1
    div(
      div("Dynamic UI:", count),
      input_dark_mode(id=paste0("a", count)),
      input_dark_mode(id=paste0("b", count)),
      input_dark_mode(id=paste0("c", count)),
    )
  })
}

shinyApp(ui, server)
```

In this screen recording, I also modified the MutationObserver's callback to do `console.log(this.id)`.

https://github.com/rstudio/bslib/assets/86978/8bc01993-a8d3-4d79-9cd3-5d7cb9b816a3



The buttons start with IDs `a1`, `b1`, `c1`. When I click on one of them, the other buttons print out their IDs to the console. So when I click on `a1`, the console prints

```
b1
c1
```

That behavior is what we'd expect.

Next, when I click the redraw button, the buttons get replaced by new ones, with IDs `a2`, `b2`, `c2`. Clicking again replaces them with `a3`, `b3`, `c3`, and so on.

I clicked on the redraw button a few times, then when I click on one of the dark mode buttons, the console prints the following, which indicates that the old ones were not getting GC'd, and their callbacks were still firing:

```
a1
b1
c1
a2
b2
c2
a3
b3
c3
a4
b4
c4
b5
c5
```


I think the way to fix this is to do the following:

- Make `_observeDocumentThemeAttribute` a static method.
- Add another static method called something like `ensureMutationObserver`, which will make sure  `_observeDocumentThemeAttribute` is called at most once.
- Add a static property which keeps track of currently live switch elements: `static activeElements: Array<BslibInputDarkMode> = [];`
- In `connectedCallback`, call `ensureMutationObserver()`, and add `this` to `activeElements`.
- In `disconnectedCallback`, remove `this` from `activeElements`.

Reply:
Thanks @wch. I was aware and Carson and I discussed this as an unlikely edge case for this particular component. 

That said, I'm happy to address this in this PR. We have an established pattern for this kind observer coordination, see the `ShinyResizeObserver`, which we implemented and use for components that are reasonably likely to be used in `renderUI()` and therefore be frequently re-connected during the lifetime of the app session.

Reply:
I just realized there's a much simpler way to deal with it than what I proposed above, and because it's simple, I think it's definitely worth doing in this PR.

In each element's `disconnectedCallback()`, we can just call its MutationObserver's `disconnect()` method. I think that's all we need to do. This will disable the callback, and break the link between the MutationObserver and `document.documentElement`. With that link broken, the element will be available for garbage collection.

I previously misunderstood how `disconnect()` worked, hence the more complicated proposal above.

Reply:
Thanks, the reprex was definitely handy! This is exactly how we handle [the IntersectionObserver used in tooltips](https://github.com/rstudio/bslib/blob/8682eb8696b65c17fe86f58190dd30f1cf0cef2f/srcts/src/components/tooltip.ts#L101-L109) (see also [my comment in that PR](https://github.com/rstudio/bslib/pull/702#discussion_r1273501487)). I'm glad we independently arrived at the same conclusion 😄

Reply:
Great, glad to hear that!